### DISCUSSION TITLE
need bufferSchema

### DISCUSSION THREAD

**Discussion by araera111** [araera111/mysql-to-zod#47] [Ideas]
There was no buffer type in zod. Therefore, it is necessary to create a bufferSchema like the following.

example
```ts
const bufferSchema = z.custom<Buffer>((value) => {
  if (Buffer.isBuffer(value)) {
    return value;
  }
  throw new Error("Not a buffer");
});
```

**Comment by araera111** [comment]
After I finished all the implementation, I noticed that there is no Buffer module in the browser.

If you use mysql2, it returns a Buffer type.
But this is not handled by the browser by default.

There is a way to convert it to an ArrayBuffer that can be used both in node.js and in the browser, using zod's transform, for example. However, some people may not like the idea of having the type changed without their permission.

I can't think of the right solution for this. I will change it to an UNKNOWN type for now and return to DISCUSSION.

**Comment by araera111** [comment]
At first, I was thinking to put the "bufferScheme" mentioned above after "import {z} from 'zod'" and change z.buffer() to bufferSchema if there is a buffer in type.

However, I noticed that the isBuffer method is not in the node.js environment.

**Comment by TonyGravagno** [comment]
This is a premature comment but I have been thinking about some of this overall ... Rather than using any primitive/defined Zod types like z.string(), z.date(), or z.buffer(), maybe a more universal approach would be to use schema named m2z.BIGINT, m2z.DATETIME, m2z.BOOLEAN, etc. Put the interpretations of these types into a separate file. This way, a developer can use this package to generate the schema structure from MySQL, but the actual interpretation of types isn't hardcoded, and can be modified by developers as they see fit.

I already do this in my code, using "schemaDateTime", "schemaBoolean", etc, so that I don't even link to a specific database. That opens yet another dimension for this package - it doesn't need to be "MySQL-to-Zod, it can be "RDB-to-Zod" or "SQL-to-Zod". A different toZod.ts module can be used for different DB platforms like Postgres, SQL Server, or Oracle.

For your purposes in this discussion, maybe you don't need to worry about the exact implementation, provide one default named m2zBuffer, m2z.buffer, or schemaBuffer. Then later, if you have ideas for different implementations, just add them. The default code can be commented out by developers so that they can use an alternative at runtime. They can even change the implementation for their runtime purpose as they wish (though this probably won't happen) by changing m2zBuffer in their application code to something like m2zBufferV1.

Example:

schemaImplementation.ts

```
export default const m2z = {
   const schemaDate = z.date()
   /** Alternates
   const schemaDate = moreComplexDateSchema()
   const schemaDate = sillyTestDateSchema()
   */
}

schema.ts
import m2z from './schemaImplementation'

export const FooSchema = z.object( {
  some_date : m2z.schemaDate
});
```

I don't think there's a need for even More sophisitication but if someone wanted to make this more complex, all of the schema can be made available, and selected at runtime via option or hardcoded parameter.

```
   const schemaDate = (which) => {
      switch(which) {
        case 0: return z.date; // always default with 0
        case 1: return moreComplexDateSchema
        case 2: return sillyTestDateSchema
      }
   } 
```

As I said, these are premature thoughts, but this is where my mind has been going as I think about solving this question about how we deal with the very real possiblity of multiple valid implementations for a given SQL Type.

This also fits with a couple of my comments about making decisions about implementations with the config.js, and a run-time config mechanism that allows an application that uses code generated here to select an application-specific choice of schema at runtime.

**Comment by araera111** [comment]
I am considering the following options.
```ts
//  toZodList
{
  zodList: [
  ["DATE", "z.date()"] or ["DATE", "z.string()"]
]
}
```

Allow the schema part corresponding to DATE to be defined by yourself as above.
If you already have your own schema or transformer, you can include it.

What about adding an import statement to the list[2] and it will be added automatically?

For example.
```ts
zodList: [
  ["DATE", "myDateSchema", "import {myDateSchema} from './globalSchema.ts' "]
]
```

```ts
import {z} from 'zod';
import {myDateSchema} from './globalSchema.ts';

const AAA = z.object({
  dateColumn: myDataSchema
});
```

**Reply by TonyGravagno** [reply]
Yes, that's exactly what I intended above with schemaImplementation.ts, which you are (better) calling globalSchema.

It doesn't matter if globalSchema exports individual Zod objects or schema, or if there is a single object like m2z that has all schema. But I was thinking that to reduce the size of the import statement, importing a single object is simpler than individually importing several type schema. Then in code it's also clear where these schema are defined : m2z.date, m2z.boolean, etc.

Yes, I like your 3-element zodList. Let's assume this utility becomes DB-to-Zod, and supports several RDBMS: MySQL, MariaDB, SQL Server, Oracle, SQLite, Postgres, etc. A developer just needs to swap a personal mysqltozod.config.js to generate schema in any of the following formats:

`["DATE", "mySqlSchema.date", "import mySqlSchema from './mySqlSchema.ts' "]`  
`["DATE", "oracleDate", "import {oracleDate} from './Schema_Oracle.ts' "]`  
`["BOOL", "schema.bool", "import mariaDB as schema from './SqlSchema.ts' "]`  

Maybe getting close to a risk of over-parameterization, this process of abstracting hard-coding out to options makes this package much more versatile. I think it's good for this software.

I will certainly help to help with implementation and to document whatever is implemented.

Some of my code might already be useful for this. If you are sure that this is the direction that you'd like to take, I will look to see if I have any code that will help.

**Reply by TonyGravagno** [reply]
Hmmm, all of this should probably be in a different thread ... oh well.

**Comment by araera111** [comment]
I think we have solved this problem, so let's close it once.
